### ISSUE TITLE
[Failsafe] Prevent topology operator to delete non-empty queues

### ISSUE THREAD

**Issue by jdfresser** [rabbitmq/messaging-topology-operator#747]
**Is your feature request related to a problem? Please describe.**
Sometimes, you could delete a queue.rabbitmq.com object by inadvertance.
For example, recently, we had an issue with our Gitops configuration that pruned everything in our clusters.
We easily recreated the topology, but the messages that where still inside the queues were lost.

**Describe the solution you'd like**
Implement a 'global' option `queue-delete-if-empty` in the topology operator.
In case of rabbitmq 'queue' objects deleted, the Queues will be deleted only if empty.

Rabbit-hole library already implements this [option](https://github.com/michaelklishin/rabbit-hole/blob/main/queues.go#L425C2-L425C9)
(Bonus: the topology operator can retry to delete it until queues are empty).

**Describe alternatives you've considered**
I've seen no similar Feature Request nor details about such a feature in the documentation.

**Comment by Zerpet** [comment]
I'm happy to see this in the Topology Operator. It will be an opt-in feature, because changing the default behaviour would be a breaking change.

I'm not sure about the global option. I think this will be more interesting if implemented as a field in the `Queue` object. There are YAML manipulation tools like Kustomize or [YTT](https://carvel.dev/ytt/) to help with adding this new field to multiple YAML manifests.

**Comment by jdfresser** [comment]
@Zerpet Agree with you, it must be an opt-in feature.

You're right, I was only considering my use-case, but it will be better to implement it as queue Queue spec item, so that users can define it by Queue  (It will also be easier to implement). . They should be mutable resources to allow users to use this feature whithout having to delete/recreate all queues. 

Should this enhancement also include the 'if-unused' option (i.e. do not delete the queue if it has consumers)?

One tricky point I didn't see at first is the incompatibility of if-empty flag with quorum queues:
```
{"error":"bad_request","reason":"cannot delete queue 'q.test' in vhost '/'. queue.delete operations with if-empty flag set are not supported by quorum queues"}
```
So, the topology operator should unset the flag if quorum queue and test the number of messages in queue before deleting it, I guess?

FYI, we do use Kustomize, but we rely on home-made python scripts to generate Topology config (and a nice mermaid schema for readability), so it'll be easy to add this field for us. We hope to simplify it by using Kluctl

**Comment by MirahImage** [comment]
Closed by #759
